File: python/leetcode/easy/house_robber.py
# https://leetcode.com/problems/house-robber/

class Solution:
    def __init__(self):
        self.cache = {}

    def rob(self, nums: List[int]) -> int:
        return max(self.total_amount(nums, 0, 0), self.total_amount(nums, 1, 0))

    def total_amount(self, nums, position, total):
        if position >= len(nums):
            self.cache[position] = total
            return total

        if position in self.cache and self.cache[position] >= total:
            return self.cache[position]

        total += nums[position]

        return max(self.total_amount(nums, position+2, total), self.total_amount(nums, position+3, total))

File: go/leetcode/easy/houseRobber.go
// https://leetcode.com/problems/house-robber/

func rob(nums []int) int {
    cache := make(map[int]int)

    totalAmount(nums, cache, 0, 0)
    totalAmount(nums, cache, 1, 0)

    result := 0
    for _, total := range(cache){
        if total > result {
            result = total
        }
    }
    return result
}

func totalAmount(nums []int, cache map[int]int, position int, total int) {
    if position >= len(nums){
        if cache[position] < total{
            cache[position] = total
        }
        return
    }
    total += nums[position]
    totalAmount(nums, cache, position+2, total)
    totalAmount(nums, cache, position+3, total)
}


